### PR TITLE
#178 support oas3 with code quality fixes

### PR DESCRIPTION
**Pull request by ruckc**

### CODE REVIEW

**Comment by ruckc** [comment]
@ioggstream - @chenjr0719 - this should fix the stale/closed #178 which needs to be reopened... I don't know why travis hasn't reported a status.

**Comment by ioggstream** [comment]
maybe we need a positive code review...

**Comment by artcg** [comment]
@ruckc Is it possible to put the black / isort reformatting in a separate pull request so this is easier to review?

@ioggstream are you satisfied that this PR is a superset of https://github.com/sanic-org/sanic-openapi/pull/179

If so, should we close https://github.com/sanic-org/sanic-openapi/pull/179 

?

**Comment by ruckc** [comment]
@artcg - i tried, but travis-ci was blocking it without all the black/isort changes.

**Comment by artcg** [comment]
Ok I have rebased this PR at https://github.com/artcg/sanic-openapi/tree/ruckc-oas3

to a version doing just the tox stuff at https://github.com/artcg/sanic-openapi/tree/ruckc-tox-only 

Lets see if it works

**Comment by artcg** [comment]
Hmm not working, do you have any idea about this error @ruckc from when you were fixing the build?

https://travis-ci.com/github/sanic-org/sanic-openapi/jobs/490299845

**Comment by ruckc** [comment]
No clue... It was extremely frustrating trying to get it all working when I tried, the tox plugins kept fighting for their own one true way to organize code, I swear I had it all working on my ci build.

**Comment by ioggstream** [comment]
@artcg can't say whether this PR is equivalent to #179 but if the `test_`s are there, I trust you and close my PR.

This PR seems to have still conflicts.

**Comment by ahopkins** [comment]
Keep me posted. If we can get some of these PRs that have been out there a while cleaned up, ping me and I will merge them. Currently v21.3 is slated to go out a week from Sunday: 2021-3-21. Maybe we should aim to get this bumped and released then.

I really am going to lean on anyone here that has the desire and ability to clean things up and get things ready to merge.

**Comment by artcg** [comment]
@ruckc if you would run 

```
git remote add artcg https://github.com/artcg/sanic-openapi
git fetch artcg
git checkout 178-support-oas3
git reset --hard artcg/ruckc-oas3
git push -f ```

On your end it will rebase this commit to the current master

**Comment by ruckc** [comment]
got the changes in, lets see travis succeed

**Comment by artcg** [comment]
ah we would still need to update the oas3 test there to co-operate with the newer sanic release but no rush to do that, can wait until everything looks good to merge

**Comment by ioggstream** [comment]
cc: @astagi

**Comment by artcg** [comment]
Update on this, I did a bit of work on `artcg:ruckc-oas3` to moving the oas3 implementation into a manner closer to the current implementation but there is still lots of work to do there, mostly:

* getting the decorators to be as compatible as possible between the implementations
* cutting down some of the bloat in the oas3 implementation which came from the external repo

But generally, the code is there, it just needs someone to connect the last pieces

I won't be finishing that however as openapi3 is not a feature I need that strongly, and I will probably be backing off this project a bit now I finished https://github.com/sanic-org/sanic-openapi/pull/208 which is the main feature I wanted in sanic-openapi, however would definitely recommend anyone who wants to finish it to use `artcg:ruckc-oas3` as a starting point, I cut down a lot from the original repo already and made sure to keep the overall spec in accordance with openapi3

**Comment by artcg** [comment]
Ok after talking to @astagi via e-mail who also helped correct some bugs here and there (thanks astagi) I was convinced to finish this,

@ruckc you can pull in the finished version via the artcg:ruck3-bugfix branch, probably will need a `reset --hard` and a force push there like before.

Long version of review:

- I was initially turned off of the new implementation (which is from zloyuser, who didnt make a PR and instead published as sanic_openapi3) due to the fact that it is a complete reimplementation with not much compatibility with the existing swagger2 implementation, but after spending some time with it it has grown on me, so I have kept some of the parts I removed earlier.
- Expanding on that point, the new implementation covers the specification much more closely than the existing (openapi2) implementation, so it will be much more easy to correct bugs, add features in future if it is ever desired

The negatives: 

- The doc.consumes, doc.produces decorators are incompatible, being replaced by the decoraters in oas3/openapi

However I feel that this is a minor point, as people can adjust their decorators if they want to upgrade to oas3 without too much effort.

To do in the future:

- Documentation would be nice, however I definitely would suggest to merge it without going crazy on the documentation because it is already turning out to be such a headache, that if we keep adding conditions to wait on it will never get merged


tl;dr usage guide:

Use exactly as before except instead of using decorators like
```python
from sanic_openapi import doc

@doc.consumes('blah')
def example():
     pass
```

Do:
```python
from sanic_openapi.oas3 import openapi

@openapi.parameter('blah')
def example():
     pass
```


I have tested it on a fairly extensive API with multiple blueprints and multiple routes for each blueprint etc, and it all seems to work OK as expected.

Other notes:

Back-compatibility is guaranteed (thanks ioggstream for initial work there) with swagger2 behaviour being accessible to the user exactly as before 


tl; dr:

lets get this merged!!!  (version with last changes on artcg:ruck3-bugfix)

@astagi might have 1-2 comments also, he was checking out some things...

**Comment by ahopkins** [comment]
Looks like a conflict to resolve here.

**Comment by artcg** [comment]
Fixed on artcg:ruck3-bugfix, if you have access to overwrite that branch to here, or otherwise ruckc can

**Comment by ahopkins** [comment]
Will do.

**Comment by ahopkins** [comment]
There are a lot of conflicts when I merge your branch in. And, when I accept all of your edits, there are some issues with missing imports. Rather than me trying to fix it and messing something up, @artcg do you want to just submit your branch as a PR and close this?

**Comment by artcg** [comment]
Sure

**Comment by ahopkins** [comment]
Closing in favor of #210
